platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

```diff
@@ -0,0 +1,350 @@
+
+*   [Update the master alias](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/update-alias.html) to the version currently aliased with the checksum.
+
+A second Github Action will be created that is [manually triggered](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/). The action will [Update and publish the function code](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionCode.html) the production lambda function.
+
+
+#### Assets
+
+Assets (images, pdfs) which are generated in Drupal will be uploaded to S3.  The content build will be updated to use the new S3 location as the `src` for the assets.  The details of this implementation will be in a different design doc and implemented outside of Small Batch Publishing.
+
+### **Code Location**
+
+The Lambda function and its supporting code will live in the `content-build` repository.   The POC code will be updated to support dev/stage/prod.
+
+The handler source code, bundling, and upload code from the POC lives in the branch `nh_preview_server_lambda` on this [PR](https://github.com/department-of-veterans-affairs/content-build/pull/177).
+
+*   Lambda function: `src/platform/lambdas/single-page-build.js`
+*   Webpack: `config/webpack.single-page-build.js`
+*   GitHub Actions: `.github/workflows/single-page-build.yml`
+
+The Lambda functions will be provisioned in the devops repo.  The POC code will be updated to support dev/stage/prod.
+
+The Drupal code will live in the `va.gov-cms` repository.
+
+
+### **Testing Plan**
+
+SBP and full build need to be tested and the output validated per content type.  The preview server currently has a `/diff` endpoint which generates a visual difference between the preview server HTML and production.  Example: [http://preview-dev.vfs.va.gov/diff?nodeId=2578](http://preview-dev.vfs.va.gov/diff?nodeId=2578). The preview will manually check for a subset of each content type before we turn on instant publishing for the content type.
+
+### **Logging**
+
+GitHub Actions will output logs for the build and versioning of the function. Log output from the function (from calling console methods) will be stored in CloudWatch.
+
+The lambda function will also communicate back to Druapl using the [Publish Batch API](#publish-batch-api) where Drupal will generate an Audit Trail of what was build by who and when.
+
+### **Debugging**
+
+*   One can start diagnosing any issues using the logs mentioned above.
+*   The Webpack build of the function can be run locally to debug issues with the build.
+*   The code should be abstracted in a way where the code can be executed outside of the lambda function.  For example, the preview server itself should be running the same code.
+*   Debugging the HTML produced and using the `/diff` endpoint which shows the HTML from the code compared to production.
```

Comment:
We use timestamps on assets as a cache buster, so you likely won't be able to get a true diff comparison here: https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/stages/build/plugins/modify-dom/process-entry-names.js

Reply:
Yeah, those and the ids added by the analytics middleware.  When we manually validate, it's easy to see them on the diff.  We don't have an automated method in mind yet.

Thanks for the heads up.